src/include/processor/operator/copy/copy_node.h

```diff
@@ -1,39 +1,108 @@
+        std::unique_ptr<PhysicalOperator> child, uint32_t id, const std::string& paramsString)
+        : Sink{std::move(resultSetDescriptor), PhysicalOperatorType::COPY_NODE, std::move(child),
               id, paramsString},
-          table{table}, nodesStatistics{nodesStatistics}, relsStore{relsStore} {}
+          localState{std::move(localState)}, sharedState{std::move(sharedState)} {}
+
+    inline void initLocalStateInternal(ResultSet* resultSet, ExecutionContext* context) override {
+        arrowDataVector = resultSet->getValueVector(localState.arrowDataPos).get();
+    }
 
-    std::unique_ptr<PhysicalOperator> clone() override {
-        return std::make_unique<CopyNode>(
-            catalog, copyDescription, table, wal, nodesStatistics, relsStore, id, paramsString);
+    inline void initGlobalStateInternal(ExecutionContext* context) override {
+        if (!isCopyAllowed()) {
+            throw common::CopyException("COPY commands can only be executed once on a table.");
+        }
+        auto nodeTableSchema = localState.catalog->getReadOnlyVersion()->getNodeTableSchema(
+            localState.table->getTableID());
+        sharedState->initialize(nodeTableSchema, localState.wal->getDirectory());
     }
 
-protected:
-    uint64_t executeInternal(
-        common::TaskScheduler* taskScheduler, ExecutionContext* executionContext) override;
+    void executeInternal(ExecutionContext* context) override;
+
+    void finalize(ExecutionContext* context) override;
+
+    inline std::unique_ptr<PhysicalOperator> clone() override {
+        return std::make_unique<CopyNode>(localState, sharedState, resultSetDescriptor->copy(),
+            children[0]->clone(), id, paramsString);
+    }
 
 private:
-    inline bool isCopyAllowed() override {
-        return nodesStatistics->getNodeStatisticsAndDeletedIDs(tableID)->getNumTuples() == 0;
+    inline bool isCopyAllowed() {
+        auto nodesStatistics = localState.table->getNodeStatisticsAndDeletedIDs();
+        return nodesStatistics->getNodeStatisticsAndDeletedIDs(localState.table->getTableID())
+                   ->getNumTuples() == 0;
     }
 
+    void flushChunksAndPopulatePKIndex(
+        const std::vector<std::unique_ptr<storage::InMemColumnChunk>>& columnChunks,
+        common::offset_t startNodeOffset, common::offset_t endNodeOffset);
+
+    template<typename T, typename... Args>
+    void appendToPKIndex(storage::InMemColumnChunk* chunk, common::offset_t startOffset,
+        uint64_t numValues, Args... args) {
+        throw common::CopyException("appendToPKIndex1 not implemented");
+    }
+
+    void populatePKIndex(storage::InMemColumnChunk* chunk, storage::InMemOverflowFile* overflowFile,
+        common::offset_t startOffset, uint64_t numValues);
+
 private:
-    storage::NodeTable* table;
-    storage::NodesStatisticsAndDeletedIDs* nodesStatistics;
-    storage::RelsStore& relsStore;
+    CopyNodeLocalState localState;
```

Comment:
unique_ptr. I think let's just have a convention to always use smart pointer. Though you can have @ray6080 to discuss this.